src/types/acls.rs

```diff
@@ -0,0 +1,372 @@
+    pub fn parse(
+        owned: Vec<u8>,
+        unsolicited: &mut mpsc::Sender<UnsolicitedResponse>,
+    ) -> Result<Self, Error> {
+        ListRightsTryBuilder {
+            data: owned,
+            rights_builder: |input| {
+                let mut rights = Vec::new();
+
+                // There should only be ONE single LISTRIGHTS response
+                parse_many_into(input, &mut rights, unsolicited, |response| match response {
+                    Response::ListRights(a) => Ok(MapOrNot::Map(InnerListRights {
+                        mailbox: a.mailbox,
+                        identifier: a.identifier,
+                        required: a.required.into(),
+                        optional: a.optional.into(),
+                    })),
+                    resp => Ok(MapOrNot::Not(resp)),
+                })?;
+
+                if rights.is_empty() {
+                    Err(Error::Parse(ParseError::Invalid(input.to_vec())))
+                } else {
+                    Ok(rights.remove(0))
+                }
+            },
+        }
+        .try_build()
+    }
+
+    /// Returns the mailbox for the rights
+    pub fn mailbox(&self) -> &str {
+        &*self.borrow_rights().mailbox
+    }
+
+    /// Returns the user identifier for the rights
+    pub fn identifier(&self) -> &str {
+        &*self.borrow_rights().identifier
+    }
+
+    /// Returns the set of rights that are always provided for this identifier
+    pub fn required(&self) -> &AclRights {
+        &self.borrow_rights().required
+    }
+
+    /// Returns the set of rights that can be granted to the identifier
+    pub fn optional(&self) -> &AclRights {
+        &self.borrow_rights().optional
+    }
+}
+
+/// Inner struct to manage storing the references for ouroboros
+#[derive(Debug, Eq, PartialEq)]
+pub(crate) struct InnerListRights<'a> {
+    /// The mailbox for the rights
+    pub(crate) mailbox: Cow<'a, str>,
+    /// The user identifier for the rights
+    pub(crate) identifier: Cow<'a, str>,
+    /// The set of rights that are always provided for this identifier
+    pub(crate) required: AclRights,
+    /// The set of rights that can be granted to the identifier
+    pub(crate) optional: AclRights,
+}
+
+/// From [section 3.8 of RFC 4313](https://datatracker.ietf.org/doc/html/rfc4314#section-3.8).
+///
+/// The MYRIGHTS response from the [Session::my_rights] IMAP command
+#[self_referencing]
+pub struct MyRights {
+    data: Vec<u8>,
+    #[borrows(data)]
+    #[covariant]
+    pub(crate) rights: InnerMyRights<'this>,
+}
+
+impl MyRights {
+    /// Parse the given input into a [MyRights] response.
+    pub fn parse(
+        owned: Vec<u8>,
+        unsolicited: &mut mpsc::Sender<UnsolicitedResponse>,
+    ) -> Result<Self, Error> {
+        MyRightsTryBuilder {
+            data: owned,
+            rights_builder: |input| {
+                let mut rights = Vec::new();
+
+                // There should only be ONE single MYRIGHTS response
```

Comment:
It does feel very weird to be using `parse_many_into` for this in so many places. I wonder if we should make a new helper, `parse_until_done`, that specifically handles the single-item case (and could internally call `parse_many_into` for now).

Reply:
I can look at that.